### ISSUE TITLE
Applying weavenet model

### ISSUE THREAD

**Issue by allchemist** [pfnet-research/chainer-chemistry#180]
Hello,
prediction with weavenet behaves strangely - my dataset has 1028 rows, but when i run prediction, the output array has only 437 rows. When i try to predict at a reduced dataset, i get even more reduced output. Other models return same prediction length with that code.

What could be the problem?

**Comment by corochann** [comment]
This is because WeaveNetPreprocessor fails to extract input feature for several molecules.

There are mainly 2 reason:
1. `num_max_atoms` need to be set.
https://github.com/pfnet-research/chainer-chemistry/blob/master/chainer_chemistry/dataset/preprocessors/weavenet_preprocessor.py#L149
We need to determine the number of max atoms beforehand for `WeaveNetPreprocessor` (this is not the case for other network, only for WeaveNet).

Current default value is set to 20, so it extracts only molecules whose number of atom is less than 20.
https://github.com/pfnet-research/chainer-chemistry/blob/master/chainer_chemistry/config.py#L7

2. WeaveNet model needs heavy man-made feature extraction, and it sometimes fail for some molecules.
 -> this is kind of a limitation.

**Comment by corochann** [comment]
btw, you can see how much feature extraction successed/failed by setting log level.
https://github.com/pfnet-research/chainer-chemistry/blob/master/examples/tox21/train_tox21.py#L38-L39

**Comment by allchemist** [comment]
Thanks. I see the reason, but such behaviour makes hard to determine the connection between input and output rows. Is it possible to return NaNs or something for failed in weavenet prediction, but return array of same size?

Also, another question - how difficult is it to deal with missed values at network training level, not at preprocessing level? It seems that you substitute missed values with '-1', at least in classification mode, while preparing data. It is an easy solution, but it means that we spoil data with third class, or imputing feature means, or etc.
In neural nets training it is possible just not to update those weights for outputs when train target is a missed values, and that would increase accuracy at multi-output tasks.

So i'm curious, how does that approach corresponds with your network structures?

**Comment by corochann** [comment]
For first question,

1. There is a on-going PR to understand which preprocessing succeeded/failed
https://github.com/pfnet-research/chainer-chemistry/pull/172

2. Parser class have an option `return_smiles` so that you can understand the smiles of extracted input features.
https://github.com/pfnet-research/chainer-chemistry/blob/master/chainer_chemistry/dataset/parsers/csv_file_parser.py#L46

For second question, 

one way is to deal with `Iterator` class to skip checking label with -1, and this is already implemented in [`BalancedSerialIterator`](https://github.com/pfnet-research/chainer-chemistry/blob/master/chainer_chemistry/iterators/balanced_serial_iterator.py). 
Usage example can be found at `tox21` example.

If you need more customized behavior, please implement your own `Iterator` class

**Comment by allchemist** [comment]
> Parser class have an option return_smiles so that you can understand the smiles of extracted input features.
Thanks, i'll try it

> one way is to deal with Iterator class to skip checking label with -1, and this is already implemented in BalancedSerialIterator.
Not sure if i understood you. I meant that if we have an example with two outputs - one present and one missing, we do not update MLP output weights for that missing output, while updating weights of other MLP output. Is that what BalancedSerialIterator allowes? I feel like it is internal network feature, not dapa preprocessing.

**Comment by corochann** [comment]
Ah in that case you can just ignore calculating "loss" for missing values, then neural network only learns for calculated loss based on non-missing values.
[`sigmoid_cross_entropy`](https://github.com/chainer/chainer/blob/master/chainer/functions/loss/sigmoid_cross_entropy.py#L122) automatically ignores label=-1.

**Comment by allchemist** [comment]
Thanks, got it. And what about regression, what could be float equivalent for missed value?

**Comment by corochann** [comment]
I think you should avoid adding as a `loss` for nan values, so that `loss.backward` propagates only loss for non-missing values.

**Comment by allchemist** [comment]
Thanks, missed values work fine now.

**Comment by corochann** [comment]
Good! We are also considering how to support missing values in library module level.
Can we close the issue?

**Comment by allchemist** [comment]
For regression i've made a modified loss function, and for classification - a postprocess_label function
https://pastebin.com/M4dQZwPt
It seems working, and if thats it, looks quite simple.

I suppose you can close, thanks for help

**Comment by allchemist** [comment]
By the way, it seems that weavenet and schnet models are not really working. Schnet model suffers exponenta overflow at any dataset with and without scaling, including your examples (softplus problem i suppose). Weavenet shows bad training accuracy - at my datasets and your qm9 example train loss is about 10 times larger than same tasks with other models (nfp, ggnn, rsgcn).

**Comment by delta2323** [comment]
Thank you for your report and sorry for our late response, if the problem of the last comment still exists, I would suggest to close this issue and create another one.
